File: backend/src/vis_platform_backend/contracts/figure_compositions.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from .common import Identifier, StrictModel
from .datasets import Dataset
from .figure_composition_content import FigureCompositionContent
from .figure_messages import FigureMessage
from .plot_runs import PlotResultSummary
from .reference_images import ReferenceImage


class CreateFigureComposition(StrictModel):
    request_id: Identifier
    content: FigureCompositionContent


class SaveFigureComposition(StrictModel):
    base_revision: int = Field(ge=1)
    content: FigureCompositionContent
    summary: str = Field(default="Edited figure", min_length=1, max_length=200)


class PanelFrame(StrictModel):
    x_mm: float
    y_mm: float
    width_mm: float
    height_mm: float


class ResolvedPanel(StrictModel):
    label: str | None = Field(description="The displayed label; null when the label is hidden.")
    frame: PanelFrame
    natural_width_mm: float
    natural_height_mm: float


class FigureCheck(StrictModel):
    code: Literal[
        "outside_margin",
        "overlap",
        "small_text",
        "low_resolution",
        "label_order",
        "unused_space",
        "empty_slot",
    ]
    severity: Literal["warning", "info"]
    message: str
    panel_ids: list[str] = Field(default_factory=list)


class FigureRenderJob(StrictModel):
    job_id: str
    panel_id: str
    status: Literal["running", "completed", "failed", "discarded"]
    width_mm: float
    height_mm: float
    error: str | None = None
    created_at: datetime


class FigureCompositionSummary(StrictModel):
    composition_id: str
    project_id: str
    title: str
    revision: int
    updated_at: datetime


class FigureCompositionDocument(FigureCompositionSummary):
    schema_version: Literal["1.0"] = "1.0"
    created_at: datetime
    content: FigureCompositionContent
    page_height_mm: float = Field(description="The resolved page height.")
    panels: dict[str, ResolvedPanel] = Field(description="Resolved geometry keyed by panel ID.")
    figures: dict[str, PlotResultSummary] = Field(
        description="Plot versions used by panels or offered as updates, keyed by version ID."
    )
    images: dict[str, ReferenceImage]
    datasets: list[Dataset] = Field(
        default_factory=list, description="The figure's datasets, in content order."
    )
    updates: dict[str, str] = Field(
        default_factory=dict, description="Newer plot versions available, keyed by panel ID."
    )
    checks: list[FigureCheck] = Field(
        default_factory=list, description="Layout and print problems; none of them block saving."
    )
    jobs: list[FigureRenderJob] = Field(
        default_factory=list, description="Recent renders of plots at their panel sizes."
    )
    messages: list[FigureMessage] = Field(
        default_factory=list, description="The figure assistant conversation, oldest first."
    )


class FigureCompositionList(StrictModel):
    schema_version: Literal["1.0"] = "1.0"
    compositions: list[FigureCompositionSummary]
    total: int
    offset: int


class FigureCompositionRevision(StrictModel):
    revision: int
    summary: str
    created_at: datetime


class FigureCompositionHistory(StrictModel):
    schema_version: Literal["1.0"] = "1.0"
    revisions: list[FigureCompositionRevision]
    total: int

File: frontend/src/api/schemas/figureCompositions.ts
import { z } from "zod";
import type { components } from "../generated/schema";
import { plotResultSchema } from "./plotRun";
import { referenceImageSchema } from "./referenceImages";

const identifier = z.string().min(1).max(160);
const millimetres = z.number().finite().min(0).max(1000);

export const figurePageSchema = z
  .object({
    width_mm: z.number().finite().min(20).max(500),
    height_mm: z.number().finite().min(20).max(1000),
    height_mode: z.enum(["fixed", "auto"]).default("auto"),
    margin_mm: z.number().finite().min(0).max(50).default(5),
  })
  .strict()
  .refine(
    (page) => 2 * page.margin_mm < Math.min(page.width_mm, page.height_mm),
    "Page margins must leave room for panels.",
  );
export const panelLabelStyleSchema = z
  .object({
    case: z.enum(["upper", "lower"]).default("upper"),
    size_pt: z.number().finite().min(5).max(24).default(10),
    bold: z.boolean().default(true),
    font_family: z
      .string()
      .regex(/^[A-Za-z0-9][A-Za-z0-9 -]*$/)
      .max(80)
      .default("Arial"),
  })
  .strict();
export const panelContentSchema = z.discriminatedUnion("type", [
  z
    .object({
      type: z.literal("plot"),
      version_id: identifier,
      source_version_id: identifier.nullable().default(null),
      ignored_version_id: identifier.nullable().default(null),
    })
    .strict(),
  z.object({ type: z.literal("image"), image_id: identifier }).strict(),
]);
export const figurePanelSchema = z
  .object({
    id: identifier,
    content: panelContentSchema,
    x_mm: millimetres,
    y_mm: millimetres,
    scale: z.number().finite().min(0.05).max(10).default(1),
    label: z
      .string()
      .min(1)
      .max(8)
      .regex(/^\S(.*\S)?$/)
      .nullable()
      .default(null),
    show_label: z.boolean().default(true),
    locked: z.boolean().default(false),
  })
  .strict();
export const figureLegendSchema = z
  .object({
    title: z.string().max(1000).default(""),
    entries: z.record(identifier, z.string().max(4000)).default({}),
  })
  .strict();
export const figureContentSchema = z
  .object({
    schema_version: z.literal("1.0").default("1.0"),
    title: z.string().min(1).max(200),
    page: figurePageSchema,
    labels: panelLabelStyleSchema.default({
      case: "upper",
      size_pt: 10,
      bold: true,
      font_family: "Arial",
    }),
    panels: z.array(figurePanelSchema).max(40).default([]),
    legend: figureLegendSchema.default({ title: "", entries: {} }),
    min_font_pt: z.number().finite().min(4).max(12).default(5),
  })
  .strict()
  .refine(
    (content) =>
      new Set(content.panels.map((panel) => panel.id)).size ===
      content.panels.length,
    "Panel IDs must be unique within a figure.",
  )
  .refine(
    (content) =>
      Object.keys(content.legend.entries).every((id) =>
        content.panels.some((panel) => panel.id === id),
      ),
    "Legend entries must refer to panels in this figure.",
  );

const frameSchema = z
  .object({
    x_mm: z.number(),
    y_mm: z.number(),
    width_mm: z.number(),
    height_mm: z.number(),
  })
  .strict();
const checkSchema = z
  .object({
    code: z.enum([
      "outside_margin",
      "overlap",
      "small_text",
      "low_resolution",
      "label_order",
      "unused_space",
    ]),
    severity: z.enum(["warning", "info"]),
    message: z.string(),
    panel_ids: z.array(z.string()).default([]),
  })
  .strict();
const jobSchema = z
  .object({
    job_id: z.string(),
    panel_id: z.string(),
    status: z.enum(["running", "completed", "failed", "discarded"]),
    width_mm: z.number(),
    height_mm: z.number(),
    error: z.string().nullable().default(null),
    created_at: z.string(),
  })
  .strict();
const summarySchema = z
  .object({
    composition_id: z.string(),
    project_id: z.string(),
    title: z.string(),
    revision: z.number().int(),
    updated_at: z.string(),
  })
  .strict();
export const figureDocumentSchema = summarySchema
  .extend({
    schema_version: z.literal("1.0"),
    created_at: z.string(),
    content: figureContentSchema,
    page_height_mm: z.number(),
    panels: z.record(
      z.string(),
      z
        .object({
          label: z.string().nullable(),
          frame: frameSchema,
          natural_width_mm: z.number(),
          natural_height_mm: z.number(),
        })
        .strict(),
    ),
    figures: z.record(z.string(), plotResultSchema),
    images: z.record(z.string(), referenceImageSchema),
    updates: z.record(z.string(), z.string()).default({}),
    checks: z.array(checkSchema).default([]),
    jobs: z.array(jobSchema).default([]),
  })
  .strict() satisfies z.ZodType<
  components["schemas"]["FigureCompositionDocument"]
>;
export const figureListSchema = z
  .object({
    schema_version: z.literal("1.0"),
    compositions: z.array(summarySchema),
    total: z.number(),
    offset: z.number(),
  })
  .strict();
export const figureHistorySchema = z
  .object({
    schema_version: z.literal("1.0"),
    revisions: z.array(
      z
        .object({
          revision: z.number(),
          summary: z.string(),
          created_at: z.string(),
        })
        .strict(),
    ),
    total: z.number(),
  })
  .strict();

export type FigureDocument = z.infer<typeof figureDocumentSchema>;
export type FigureContent = z.infer<typeof figureContentSchema>;
export type FigurePage = z.infer<typeof figurePageSchema>;
export type FigurePanel = z.infer<typeof figurePanelSchema>;
export type FigureLegend = z.infer<typeof figureLegendSchema>;
export type PanelLabelStyle = z.infer<typeof panelLabelStyleSchema>;
export type PanelFrame = z.infer<typeof frameSchema>;
export type FigureCheck = z.infer<typeof checkSchema>;
export type FigureRenderJob = z.infer<typeof jobSchema>;
export type FigureArrangement =
  components["schemas"]["ArrangeRequest"]["arrangement"];
export type PanelGeometry = Pick<FigurePanel, "x_mm" | "y_mm" | "scale">;
export type FigureOperation =
  components["schemas"]["FigureOperationsRequest"]["operations"][number];
